{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "3e68fba1",
   "metadata": {},
   "outputs": [],
   "source": [
    "import subprocess\n",
    "from Bio import Entrez\n",
    "from Bio import SeqIO\n",
    "import pandas as pd\n",
    "Entrez.email = \"o.william.white@gmail.com\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "1cd37195",
   "metadata": {},
   "outputs": [],
   "source": [
    "# input parameters\n",
    "reference_accession = \"NC_000932.1\"\n",
    "query_accession = \"NC_000932.1\"\n",
    "query_circular = True"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "18223466",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Note 2 sequences have been extracted under the name trnS\n",
      "Note 2 sequences have been extracted under the name trnG\n",
      "Note 3 sequences have been extracted under the name trnS\n",
      "Ignoring duplicate annotation for rps12\n",
      "Note 2 sequences have been extracted under the name trnL\n",
      "Note 2 sequences have been extracted under the name trnV\n",
      "Note 2 sequences have been extracted under the name trnI\n",
      "Note 2 sequences have been extracted under the name trnR\n",
      "Note 3 sequences have been extracted under the name trnL\n",
      "Note 2 sequences have been extracted under the name ycf1\n",
      "Ignoring duplicate annotation for trnN\n",
      "Note 3 sequences have been extracted under the name trnR\n",
      "Ignoring duplicate annotation for trnA\n",
      "Note 3 sequences have been extracted under the name trnI\n",
      "Ignoring duplicate annotation for rps7\n",
      "Ignoring duplicate annotation for ndhB\n",
      "Note 4 sequences have been extracted under the name trnL\n",
      "Ignoring duplicate annotation for ycf2\n",
      "Ignoring duplicate annotation for trnI\n",
      "Ignoring duplicate annotation for rpl23\n",
      "Ignoring duplicate annotation for rpl2\n"
     ]
    }
   ],
   "source": [
    "## download genbank of reference data and write annotated genes to fasta\n",
    "\n",
    "# efetch accession\n",
    "handle = Entrez.efetch(db=\"nucleotide\", id=reference_accession, rettype=\"gb\", retmode=\"text\")\n",
    "reference_record = SeqIO.read(handle, \"gb\")\n",
    "handle.close()\n",
    "\n",
    "# list and dictionary to store annotations and counts respectively\n",
    "list_coding, list_non_coding, dict_genes, dict_counts = [], [], {}, {}\n",
    "\n",
    "# loop through sequence feature and append to list\n",
    "for feature in reference_record.features:\n",
    "    # cds features\n",
    "    if (feature.type in [\"CDS\", \"tRNA\", \"rRNA\"]) and (\"gene\" in feature.qualifiers.keys()):\n",
    "        # get annotation info\n",
    "        gene_name = str(feature.qualifiers[\"gene\"][0])\n",
    "        gene_seq = feature.extract(reference_record)\n",
    "        gene_seq.description = \"\"\n",
    "        if dict_genes.get(gene_name) == None:\n",
    "            dict_genes[gene_name] = str(gene_seq.seq)\n",
    "            dict_counts[gene_name] = 1\n",
    "        elif dict_genes[gene_name] == str(gene_seq.seq): \n",
    "            print(f\"Ignoring duplicate annotation for {gene_name}\")\n",
    "            continue\n",
    "        elif dict_genes[gene_name] != str(gene_seq.seq):\n",
    "            dict_counts[gene_name] += 1\n",
    "            print(f\"Note {dict_counts[gene_name]} sequences have been extracted under the name {gene_name}\")\n",
    "        if feature.type == \"CDS\": \n",
    "            gene_seq.id = f\"cds-blatx_{gene_name}_{dict_counts[gene_name]};gbkey={feature.type};gene={gene_name}\"\n",
    "            list_coding.append(gene_seq)\n",
    "        elif (feature.type == \"tRNA\") or (feature.type == \"rRNA\"):\n",
    "            gene_seq.id = f\"cds-blatn_{gene_name}_{dict_counts[gene_name]};gbkey={feature.type};gene={gene_name}\"\n",
    "            list_non_coding.append(gene_seq)\n",
    "\n",
    "# write reference fasta\n",
    "with open(\"reference_coding.fasta\", \"w\") as coding_fas:\n",
    "    SeqIO.write(list_coding, coding_fas, \"fasta\")\n",
    "\n",
    "with open(\"reference_non_coding.fasta\", \"w\") as non_coding_fas:\n",
    "    SeqIO.write(list_non_coding, non_coding_fas, \"fasta\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "bd0a1d1f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Treating query as circular\n"
     ]
    }
   ],
   "source": [
    "## download query accession and write to fasta\n",
    "\n",
    "# efetch accession\n",
    "handle = Entrez.efetch(db=\"nucleotide\", id=query_accession, rettype=\"fasta\", retmode=\"text\")\n",
    "query_record = SeqIO.read(handle, \"fasta\")\n",
    "handle.close()\n",
    "\n",
    "# if the query is circular, append the first 10Kb to the end of the sequence query\n",
    "if query_circular:\n",
    "    print(\"Treating query as circular\")\n",
    "    query_record.seq = query_record.seq + query_record.seq[0:10000]\n",
    "\n",
    "# write reference fasta\n",
    "with open(\"query.fasta\", \"w\") as que:\n",
    "    SeqIO.write(query_record, que, \"fasta\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "31f18e06",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Loaded 69111 letters in 79 sequences\n",
      "Blatx 79 sequences in database, 1 files in query\n",
      "Loaded 6829 letters in 35 sequences\n",
      "Searched 164478 bases in 1 sequences\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "CompletedProcess(args=['blat', 'reference_non_coding.fasta', 'query.fasta', '-t=dna', '-q=dna', '-out=blast8', 'output_non_coding.txt'], returncode=0)"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# run blat\n",
    "subprocess.run([\"blat\", \"reference_coding.fasta\",      \"query.fasta\", \"-t=dnax\", \"-q=dnax\", \"-out=blast8\", \"output_coding.txt\"])\n",
    "subprocess.run([\"blat\", \"reference_non_coding.fasta\",  \"query.fasta\", \"-t=dna\",  \"-q=dna\",  \"-out=blast8\", \"output_non_coding.txt\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "bc1533eb",
   "metadata": {},
   "outputs": [],
   "source": [
    "# read blat output into pandas dataframes\n",
    "blast8_columns = [\"qseqid\", \"sseqid\", \"pident\", \"length\", \"mismatch\", \"gapopen\", \"qstart\", \"qend\", \"sstart\", \"send\", \"evalue\", \"bitscore\"]\n",
    "blat_coding     = pd.read_csv(\"output_coding.txt\",     sep=\"\\t\", names=blast8_columns)\n",
    "blat_non_coding = pd.read_csv(\"output_non_coding.txt\", sep=\"\\t\", names=blast8_columns)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "73c057a5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>qseqid</th>\n",
       "      <th>sseqid</th>\n",
       "      <th>pident</th>\n",
       "      <th>length</th>\n",
       "      <th>mismatch</th>\n",
       "      <th>gapopen</th>\n",
       "      <th>qstart</th>\n",
       "      <th>qend</th>\n",
       "      <th>sstart</th>\n",
       "      <th>send</th>\n",
       "      <th>evalue</th>\n",
       "      <th>bitscore</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NC_000932.1</td>\n",
       "      <td>cds-blatx_ycf2_1;gbkey=CDS;gene=ycf2</td>\n",
       "      <td>100.00</td>\n",
       "      <td>6885</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>86474</td>\n",
       "      <td>93358</td>\n",
       "      <td>1</td>\n",
       "      <td>6885</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>13318.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NC_000932.1</td>\n",
       "      <td>cds-blatx_ycf2_1;gbkey=CDS;gene=ycf2</td>\n",
       "      <td>100.00</td>\n",
       "      <td>6885</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>145291</td>\n",
       "      <td>152175</td>\n",
       "      <td>6885</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>13318.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NC_000932.1</td>\n",
       "      <td>cds-blatx_ycf2_1;gbkey=CDS;gene=ycf2</td>\n",
       "      <td>84.54</td>\n",
       "      <td>97</td>\n",
       "      <td>12</td>\n",
       "      <td>1</td>\n",
       "      <td>89060</td>\n",
       "      <td>89156</td>\n",
       "      <td>2611</td>\n",
       "      <td>2704</td>\n",
       "      <td>1.600000e-31</td>\n",
       "      <td>134.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NC_000932.1</td>\n",
       "      <td>cds-blatx_ycf2_1;gbkey=CDS;gene=ycf2</td>\n",
       "      <td>84.54</td>\n",
       "      <td>97</td>\n",
       "      <td>12</td>\n",
       "      <td>1</td>\n",
       "      <td>149493</td>\n",
       "      <td>149589</td>\n",
       "      <td>2704</td>\n",
       "      <td>2611</td>\n",
       "      <td>1.600000e-31</td>\n",
       "      <td>134.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NC_000932.1</td>\n",
       "      <td>cds-blatx_ycf2_1;gbkey=CDS;gene=ycf2</td>\n",
       "      <td>86.81</td>\n",
       "      <td>91</td>\n",
       "      <td>9</td>\n",
       "      <td>1</td>\n",
       "      <td>89083</td>\n",
       "      <td>89170</td>\n",
       "      <td>2586</td>\n",
       "      <td>2676</td>\n",
       "      <td>5.400000e-31</td>\n",
       "      <td>132.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>242</th>\n",
       "      <td>NC_000932.1</td>\n",
       "      <td>cds-blatx_psbI_1;gbkey=CDS;gene=psbI</td>\n",
       "      <td>100.00</td>\n",
       "      <td>111</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>162061</td>\n",
       "      <td>162171</td>\n",
       "      <td>1</td>\n",
       "      <td>111</td>\n",
       "      <td>1.100000e-55</td>\n",
       "      <td>214.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>243</th>\n",
       "      <td>NC_000932.1</td>\n",
       "      <td>cds-blatx_psbM_1;gbkey=CDS;gene=psbM</td>\n",
       "      <td>100.00</td>\n",
       "      <td>105</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>28707</td>\n",
       "      <td>28811</td>\n",
       "      <td>105</td>\n",
       "      <td>1</td>\n",
       "      <td>5.400000e-52</td>\n",
       "      <td>202.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>244</th>\n",
       "      <td>NC_000932.1</td>\n",
       "      <td>cds-blatx_psbT_1;gbkey=CDS;gene=psbT</td>\n",
       "      <td>100.00</td>\n",
       "      <td>102</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>74082</td>\n",
       "      <td>74183</td>\n",
       "      <td>1</td>\n",
       "      <td>102</td>\n",
       "      <td>3.400000e-50</td>\n",
       "      <td>196.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>245</th>\n",
       "      <td>NC_000932.1</td>\n",
       "      <td>cds-blatx_petL_1;gbkey=CDS;gene=petL</td>\n",
       "      <td>100.00</td>\n",
       "      <td>96</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>65712</td>\n",
       "      <td>65807</td>\n",
       "      <td>1</td>\n",
       "      <td>96</td>\n",
       "      <td>7.000000e-47</td>\n",
       "      <td>185.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>246</th>\n",
       "      <td>NC_000932.1</td>\n",
       "      <td>cds-blatx_petN_1;gbkey=CDS;gene=petN</td>\n",
       "      <td>100.00</td>\n",
       "      <td>90</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>28089</td>\n",
       "      <td>28178</td>\n",
       "      <td>1</td>\n",
       "      <td>90</td>\n",
       "      <td>5.900000e-44</td>\n",
       "      <td>175.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>247 rows × 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          qseqid                                sseqid  pident  length  \\\n",
       "0    NC_000932.1  cds-blatx_ycf2_1;gbkey=CDS;gene=ycf2  100.00    6885   \n",
       "1    NC_000932.1  cds-blatx_ycf2_1;gbkey=CDS;gene=ycf2  100.00    6885   \n",
       "2    NC_000932.1  cds-blatx_ycf2_1;gbkey=CDS;gene=ycf2   84.54      97   \n",
       "3    NC_000932.1  cds-blatx_ycf2_1;gbkey=CDS;gene=ycf2   84.54      97   \n",
       "4    NC_000932.1  cds-blatx_ycf2_1;gbkey=CDS;gene=ycf2   86.81      91   \n",
       "..           ...                                   ...     ...     ...   \n",
       "242  NC_000932.1  cds-blatx_psbI_1;gbkey=CDS;gene=psbI  100.00     111   \n",
       "243  NC_000932.1  cds-blatx_psbM_1;gbkey=CDS;gene=psbM  100.00     105   \n",
       "244  NC_000932.1  cds-blatx_psbT_1;gbkey=CDS;gene=psbT  100.00     102   \n",
       "245  NC_000932.1  cds-blatx_petL_1;gbkey=CDS;gene=petL  100.00      96   \n",
       "246  NC_000932.1  cds-blatx_petN_1;gbkey=CDS;gene=petN  100.00      90   \n",
       "\n",
       "     mismatch  gapopen  qstart    qend  sstart  send        evalue  bitscore  \n",
       "0           0        0   86474   93358       1  6885  0.000000e+00   13318.0  \n",
       "1           0        0  145291  152175    6885     1  0.000000e+00   13318.0  \n",
       "2          12        1   89060   89156    2611  2704  1.600000e-31     134.0  \n",
       "3          12        1  149493  149589    2704  2611  1.600000e-31     134.0  \n",
       "4           9        1   89083   89170    2586  2676  5.400000e-31     132.0  \n",
       "..        ...      ...     ...     ...     ...   ...           ...       ...  \n",
       "242         0        0  162061  162171       1   111  1.100000e-55     214.0  \n",
       "243         0        0   28707   28811     105     1  5.400000e-52     202.0  \n",
       "244         0        0   74082   74183       1   102  3.400000e-50     196.0  \n",
       "245         0        0   65712   65807       1    96  7.000000e-47     185.0  \n",
       "246         0        0   28089   28178       1    90  5.900000e-44     175.0  \n",
       "\n",
       "[247 rows x 12 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# check \n",
    "blat_coding"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "1eec8f6b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>qseqid</th>\n",
       "      <th>sseqid</th>\n",
       "      <th>pident</th>\n",
       "      <th>length</th>\n",
       "      <th>mismatch</th>\n",
       "      <th>gapopen</th>\n",
       "      <th>qstart</th>\n",
       "      <th>qend</th>\n",
       "      <th>sstart</th>\n",
       "      <th>send</th>\n",
       "      <th>evalue</th>\n",
       "      <th>bitscore</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NC_000932.1</td>\n",
       "      <td>cds-blatn_rrn23S_1;gbkey=rRNA;gene=rrn23S</td>\n",
       "      <td>100.0</td>\n",
       "      <td>2810</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>104691</td>\n",
       "      <td>107500</td>\n",
       "      <td>1</td>\n",
       "      <td>2810</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>5528.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NC_000932.1</td>\n",
       "      <td>cds-blatn_rrn23S_1;gbkey=rRNA;gene=rrn23S</td>\n",
       "      <td>100.0</td>\n",
       "      <td>2810</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>131149</td>\n",
       "      <td>133958</td>\n",
       "      <td>2810</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>5528.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NC_000932.1</td>\n",
       "      <td>cds-blatn_rrn16S_1;gbkey=rRNA;gene=rrn16S</td>\n",
       "      <td>100.0</td>\n",
       "      <td>1491</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>101012</td>\n",
       "      <td>102502</td>\n",
       "      <td>1</td>\n",
       "      <td>1491</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>2937.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NC_000932.1</td>\n",
       "      <td>cds-blatn_rrn16S_1;gbkey=rRNA;gene=rrn16S</td>\n",
       "      <td>100.0</td>\n",
       "      <td>1491</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>136147</td>\n",
       "      <td>137637</td>\n",
       "      <td>1491</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>2937.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NC_000932.1</td>\n",
       "      <td>cds-blatn_rrn4.5S_1;gbkey=rRNA;gene=rrn4.5S</td>\n",
       "      <td>100.0</td>\n",
       "      <td>103</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>107599</td>\n",
       "      <td>107701</td>\n",
       "      <td>1</td>\n",
       "      <td>103</td>\n",
       "      <td>5.600000e-52</td>\n",
       "      <td>202.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>71</th>\n",
       "      <td>NC_000932.1</td>\n",
       "      <td>cds-blatn_trnI_2;gbkey=tRNA;gene=trnI</td>\n",
       "      <td>100.0</td>\n",
       "      <td>35</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>135048</td>\n",
       "      <td>135082</td>\n",
       "      <td>72</td>\n",
       "      <td>38</td>\n",
       "      <td>4.500000e-12</td>\n",
       "      <td>69.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>72</th>\n",
       "      <td>NC_000932.1</td>\n",
       "      <td>cds-blatn_trnK_1;gbkey=tRNA;gene=trnK</td>\n",
       "      <td>100.0</td>\n",
       "      <td>37</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>4311</td>\n",
       "      <td>4347</td>\n",
       "      <td>37</td>\n",
       "      <td>1</td>\n",
       "      <td>5.700000e-13</td>\n",
       "      <td>72.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>73</th>\n",
       "      <td>NC_000932.1</td>\n",
       "      <td>cds-blatn_trnK_1;gbkey=tRNA;gene=trnK</td>\n",
       "      <td>100.0</td>\n",
       "      <td>37</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>158789</td>\n",
       "      <td>158825</td>\n",
       "      <td>37</td>\n",
       "      <td>1</td>\n",
       "      <td>5.700000e-13</td>\n",
       "      <td>72.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>74</th>\n",
       "      <td>NC_000932.1</td>\n",
       "      <td>cds-blatn_trnK_1;gbkey=tRNA;gene=trnK</td>\n",
       "      <td>100.0</td>\n",
       "      <td>35</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1717</td>\n",
       "      <td>1751</td>\n",
       "      <td>72</td>\n",
       "      <td>38</td>\n",
       "      <td>3.500000e-12</td>\n",
       "      <td>70.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75</th>\n",
       "      <td>NC_000932.1</td>\n",
       "      <td>cds-blatn_trnK_1;gbkey=tRNA;gene=trnK</td>\n",
       "      <td>100.0</td>\n",
       "      <td>35</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>156195</td>\n",
       "      <td>156229</td>\n",
       "      <td>72</td>\n",
       "      <td>38</td>\n",
       "      <td>3.500000e-12</td>\n",
       "      <td>70.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>76 rows × 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         qseqid                                       sseqid  pident  length  \\\n",
       "0   NC_000932.1    cds-blatn_rrn23S_1;gbkey=rRNA;gene=rrn23S   100.0    2810   \n",
       "1   NC_000932.1    cds-blatn_rrn23S_1;gbkey=rRNA;gene=rrn23S   100.0    2810   \n",
       "2   NC_000932.1    cds-blatn_rrn16S_1;gbkey=rRNA;gene=rrn16S   100.0    1491   \n",
       "3   NC_000932.1    cds-blatn_rrn16S_1;gbkey=rRNA;gene=rrn16S   100.0    1491   \n",
       "4   NC_000932.1  cds-blatn_rrn4.5S_1;gbkey=rRNA;gene=rrn4.5S   100.0     103   \n",
       "..          ...                                          ...     ...     ...   \n",
       "71  NC_000932.1        cds-blatn_trnI_2;gbkey=tRNA;gene=trnI   100.0      35   \n",
       "72  NC_000932.1        cds-blatn_trnK_1;gbkey=tRNA;gene=trnK   100.0      37   \n",
       "73  NC_000932.1        cds-blatn_trnK_1;gbkey=tRNA;gene=trnK   100.0      37   \n",
       "74  NC_000932.1        cds-blatn_trnK_1;gbkey=tRNA;gene=trnK   100.0      35   \n",
       "75  NC_000932.1        cds-blatn_trnK_1;gbkey=tRNA;gene=trnK   100.0      35   \n",
       "\n",
       "    mismatch  gapopen  qstart    qend  sstart  send        evalue  bitscore  \n",
       "0          0        0  104691  107500       1  2810  0.000000e+00    5528.0  \n",
       "1          0        0  131149  133958    2810     1  0.000000e+00    5528.0  \n",
       "2          0        0  101012  102502       1  1491  0.000000e+00    2937.0  \n",
       "3          0        0  136147  137637    1491     1  0.000000e+00    2937.0  \n",
       "4          0        0  107599  107701       1   103  5.600000e-52     202.0  \n",
       "..       ...      ...     ...     ...     ...   ...           ...       ...  \n",
       "71         0        0  135048  135082      72    38  4.500000e-12      69.0  \n",
       "72         0        0    4311    4347      37     1  5.700000e-13      72.0  \n",
       "73         0        0  158789  158825      37     1  5.700000e-13      72.0  \n",
       "74         0        0    1717    1751      72    38  3.500000e-12      70.0  \n",
       "75         0        0  156195  156229      72    38  3.500000e-12      70.0  \n",
       "\n",
       "[76 rows x 12 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "blat_non_coding"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "06195511",
   "metadata": {},
   "outputs": [],
   "source": [
    "# select top hits based on bitscores\n",
    "blat_coding_top     = blat_coding.loc[blat_coding.groupby('sseqid')['bitscore'].idxmax()]\n",
    "blat_non_coding_top = blat_non_coding.loc[blat_non_coding.groupby('sseqid')['bitscore'].idxmax()]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "314e3701",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index</th>\n",
       "      <th>qseqid</th>\n",
       "      <th>sseqid</th>\n",
       "      <th>pident</th>\n",
       "      <th>length</th>\n",
       "      <th>mismatch</th>\n",
       "      <th>gapopen</th>\n",
       "      <th>qstart</th>\n",
       "      <th>qend</th>\n",
       "      <th>sstart</th>\n",
       "      <th>send</th>\n",
       "      <th>evalue</th>\n",
       "      <th>bitscore</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>37</td>\n",
       "      <td>NC_000932.1</td>\n",
       "      <td>cds-blatn_trnH_1;gbkey=tRNA;gene=trnH</td>\n",
       "      <td>100.00</td>\n",
       "      <td>73</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>76</td>\n",
       "      <td>73</td>\n",
       "      <td>1</td>\n",
       "      <td>2.100000e-34</td>\n",
       "      <td>143.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>36</td>\n",
       "      <td>NC_000932.1</td>\n",
       "      <td>cds-blatx_psbA_1;gbkey=CDS;gene=psbA</td>\n",
       "      <td>100.00</td>\n",
       "      <td>1062</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>383</td>\n",
       "      <td>1444</td>\n",
       "      <td>1062</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>2062.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>27</td>\n",
       "      <td>NC_000932.1</td>\n",
       "      <td>cds-blatx_matK_1;gbkey=CDS;gene=matK</td>\n",
       "      <td>100.00</td>\n",
       "      <td>1515</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2056</td>\n",
       "      <td>3570</td>\n",
       "      <td>1515</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>2914.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>72</td>\n",
       "      <td>NC_000932.1</td>\n",
       "      <td>cds-blatn_trnK_1;gbkey=tRNA;gene=trnK</td>\n",
       "      <td>100.00</td>\n",
       "      <td>37</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>4311</td>\n",
       "      <td>4347</td>\n",
       "      <td>37</td>\n",
       "      <td>1</td>\n",
       "      <td>5.700000e-13</td>\n",
       "      <td>72.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>100</td>\n",
       "      <td>NC_000932.1</td>\n",
       "      <td>cds-blatx_rps16_1;gbkey=CDS;gene=rps16</td>\n",
       "      <td>100.00</td>\n",
       "      <td>204</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>5084</td>\n",
       "      <td>5287</td>\n",
       "      <td>240</td>\n",
       "      <td>37</td>\n",
       "      <td>5.600000e-110</td>\n",
       "      <td>394.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>109</th>\n",
       "      <td>64</td>\n",
       "      <td>NC_000932.1</td>\n",
       "      <td>cds-blatx_rpl2_1;gbkey=CDS;gene=rpl2</td>\n",
       "      <td>99.55</td>\n",
       "      <td>440</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>153873</td>\n",
       "      <td>154312</td>\n",
       "      <td>386</td>\n",
       "      <td>825</td>\n",
       "      <td>5.300000e-248</td>\n",
       "      <td>853.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>110</th>\n",
       "      <td>106</td>\n",
       "      <td>NC_000932.1</td>\n",
       "      <td>cds-blatx_psbK_1;gbkey=CDS;gene=psbK</td>\n",
       "      <td>100.00</td>\n",
       "      <td>186</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>161495</td>\n",
       "      <td>161680</td>\n",
       "      <td>1</td>\n",
       "      <td>186</td>\n",
       "      <td>2.600000e-99</td>\n",
       "      <td>359.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>111</th>\n",
       "      <td>117</td>\n",
       "      <td>NC_000932.1</td>\n",
       "      <td>cds-blatx_psbI_1;gbkey=CDS;gene=psbI</td>\n",
       "      <td>100.00</td>\n",
       "      <td>111</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>162061</td>\n",
       "      <td>162171</td>\n",
       "      <td>1</td>\n",
       "      <td>111</td>\n",
       "      <td>1.100000e-55</td>\n",
       "      <td>214.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>112</th>\n",
       "      <td>9</td>\n",
       "      <td>NC_000932.1</td>\n",
       "      <td>cds-blatn_trnS_1;gbkey=tRNA;gene=trnS</td>\n",
       "      <td>100.00</td>\n",
       "      <td>88</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>162263</td>\n",
       "      <td>162350</td>\n",
       "      <td>1</td>\n",
       "      <td>88</td>\n",
       "      <td>4.700000e-43</td>\n",
       "      <td>172.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>113</th>\n",
       "      <td>49</td>\n",
       "      <td>NC_000932.1</td>\n",
       "      <td>cds-blatn_trnR_1;gbkey=tRNA;gene=trnR</td>\n",
       "      <td>100.00</td>\n",
       "      <td>72</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>164068</td>\n",
       "      <td>164139</td>\n",
       "      <td>1</td>\n",
       "      <td>72</td>\n",
       "      <td>2.400000e-33</td>\n",
       "      <td>140.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>114 rows × 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     index       qseqid                                  sseqid  pident  \\\n",
       "0       37  NC_000932.1   cds-blatn_trnH_1;gbkey=tRNA;gene=trnH  100.00   \n",
       "1       36  NC_000932.1    cds-blatx_psbA_1;gbkey=CDS;gene=psbA  100.00   \n",
       "2       27  NC_000932.1    cds-blatx_matK_1;gbkey=CDS;gene=matK  100.00   \n",
       "3       72  NC_000932.1   cds-blatn_trnK_1;gbkey=tRNA;gene=trnK  100.00   \n",
       "4      100  NC_000932.1  cds-blatx_rps16_1;gbkey=CDS;gene=rps16  100.00   \n",
       "..     ...          ...                                     ...     ...   \n",
       "109     64  NC_000932.1    cds-blatx_rpl2_1;gbkey=CDS;gene=rpl2   99.55   \n",
       "110    106  NC_000932.1    cds-blatx_psbK_1;gbkey=CDS;gene=psbK  100.00   \n",
       "111    117  NC_000932.1    cds-blatx_psbI_1;gbkey=CDS;gene=psbI  100.00   \n",
       "112      9  NC_000932.1   cds-blatn_trnS_1;gbkey=tRNA;gene=trnS  100.00   \n",
       "113     49  NC_000932.1   cds-blatn_trnR_1;gbkey=tRNA;gene=trnR  100.00   \n",
       "\n",
       "     length  mismatch  gapopen  qstart    qend  sstart  send         evalue  \\\n",
       "0        73         0        0       4      76      73     1   2.100000e-34   \n",
       "1      1062         0        0     383    1444    1062     1   0.000000e+00   \n",
       "2      1515         0        0    2056    3570    1515     1   0.000000e+00   \n",
       "3        37         0        0    4311    4347      37     1   5.700000e-13   \n",
       "4       204         0        0    5084    5287     240    37  5.600000e-110   \n",
       "..      ...       ...      ...     ...     ...     ...   ...            ...   \n",
       "109     440         2        0  153873  154312     386   825  5.300000e-248   \n",
       "110     186         0        0  161495  161680       1   186   2.600000e-99   \n",
       "111     111         0        0  162061  162171       1   111   1.100000e-55   \n",
       "112      88         0        0  162263  162350       1    88   4.700000e-43   \n",
       "113      72         0        0  164068  164139       1    72   2.400000e-33   \n",
       "\n",
       "     bitscore  \n",
       "0       143.0  \n",
       "1      2062.0  \n",
       "2      2914.0  \n",
       "3        72.0  \n",
       "4       394.0  \n",
       "..        ...  \n",
       "109     853.0  \n",
       "110     359.0  \n",
       "111     214.0  \n",
       "112     172.0  \n",
       "113     140.0  \n",
       "\n",
       "[114 rows x 13 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# concat blat outputs\n",
    "dat = pd.concat([blat_coding_top, blat_non_coding_top], axis=0)\n",
    "dat = dat.sort_values(by=\"qstart\").reset_index()\n",
    "dat"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "51b6ffc3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index</th>\n",
       "      <th>qseqid</th>\n",
       "      <th>sseqid</th>\n",
       "      <th>pident</th>\n",
       "      <th>length</th>\n",
       "      <th>mismatch</th>\n",
       "      <th>gapopen</th>\n",
       "      <th>qstart</th>\n",
       "      <th>qend</th>\n",
       "      <th>sstart</th>\n",
       "      <th>send</th>\n",
       "      <th>evalue</th>\n",
       "      <th>bitscore</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>27</td>\n",
       "      <td>NC_000932.1</td>\n",
       "      <td>cds-blatx_matK_1;gbkey=CDS;gene=matK</td>\n",
       "      <td>100.0</td>\n",
       "      <td>1515</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2056</td>\n",
       "      <td>3570</td>\n",
       "      <td>1515</td>\n",
       "      <td>1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2914.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   index       qseqid                                sseqid  pident  length  \\\n",
       "2     27  NC_000932.1  cds-blatx_matK_1;gbkey=CDS;gene=matK   100.0    1515   \n",
       "\n",
       "   mismatch  gapopen  qstart  qend  sstart  send  evalue  bitscore  \n",
       "2         0        0    2056  3570    1515     1     0.0    2914.0  "
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dat [ dat[\"sseqid\"] == \"cds-blatx_matK_1;gbkey=CDS;gene=matK\"]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "47e76567",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index</th>\n",
       "      <th>qseqid</th>\n",
       "      <th>sseqid</th>\n",
       "      <th>pident</th>\n",
       "      <th>length</th>\n",
       "      <th>mismatch</th>\n",
       "      <th>gapopen</th>\n",
       "      <th>qstart</th>\n",
       "      <th>qend</th>\n",
       "      <th>sstart</th>\n",
       "      <th>send</th>\n",
       "      <th>evalue</th>\n",
       "      <th>bitscore</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>107</th>\n",
       "      <td>34</td>\n",
       "      <td>NC_000932.1</td>\n",
       "      <td>cds-blatn_trnI_1;gbkey=tRNA;gene=trnI</td>\n",
       "      <td>100.0</td>\n",
       "      <td>74</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>152264</td>\n",
       "      <td>152337</td>\n",
       "      <td>1</td>\n",
       "      <td>74</td>\n",
       "      <td>1.100000e-34</td>\n",
       "      <td>144.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     index       qseqid                                 sseqid  pident  \\\n",
       "107     34  NC_000932.1  cds-blatn_trnI_1;gbkey=tRNA;gene=trnI   100.0   \n",
       "\n",
       "     length  mismatch  gapopen  qstart    qend  sstart  send        evalue  \\\n",
       "107      74         0        0  152264  152337       1    74  1.100000e-34   \n",
       "\n",
       "     bitscore  \n",
       "107     144.0  "
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dat [ dat[\"sseqid\"] == \"cds-blatn_trnI_1;gbkey=tRNA;gene=trnI\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "ee816092",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[SeqFeature(FeatureLocation(ExactPosition(0), ExactPosition(154478), strand=1), type='source'),\n",
       " SeqFeature(CompoundLocation([FeatureLocation(ExactPosition(69610), ExactPosition(69724), strand=-1), FeatureLocation(ExactPosition(97998), ExactPosition(98793), strand=-1)], 'join'), type='gene', location_operator='join'),\n",
       " SeqFeature(CompoundLocation([FeatureLocation(ExactPosition(69610), ExactPosition(69724), strand=-1), FeatureLocation(ExactPosition(98561), ExactPosition(98793), strand=-1), FeatureLocation(ExactPosition(97998), ExactPosition(98024), strand=-1)], 'join'), type='CDS', location_operator='join'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(3), ExactPosition(76), strand=-1), type='gene'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(3), ExactPosition(76), strand=-1), type='tRNA'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(382), ExactPosition(1444), strand=-1), type='gene'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(382), ExactPosition(1444), strand=-1), type='CDS'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(1716), ExactPosition(4347), strand=-1), type='gene'),\n",
       " SeqFeature(CompoundLocation([FeatureLocation(ExactPosition(4310), ExactPosition(4347), strand=-1), FeatureLocation(ExactPosition(1716), ExactPosition(1751), strand=-1)], 'join'), type='tRNA', location_operator='join'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(2055), ExactPosition(3570), strand=-1), type='gene'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(2055), ExactPosition(3570), strand=-1), type='CDS'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(5083), ExactPosition(6188), strand=-1), type='gene'),\n",
       " SeqFeature(CompoundLocation([FeatureLocation(ExactPosition(6148), ExactPosition(6188), strand=-1), FeatureLocation(ExactPosition(5083), ExactPosition(5283), strand=-1)], 'join'), type='CDS', location_operator='join'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(6615), ExactPosition(6687), strand=-1), type='gene'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(6615), ExactPosition(6687), strand=-1), type='tRNA'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(7016), ExactPosition(7202), strand=1), type='gene'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(7016), ExactPosition(7202), strand=1), type='CDS'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(7582), ExactPosition(7693), strand=1), type='gene'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(7582), ExactPosition(7693), strand=1), type='CDS'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(7784), ExactPosition(7872), strand=1), type='gene'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(7784), ExactPosition(7872), strand=1), type='tRNA'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(8645), ExactPosition(9431), strand=1), type='gene'),\n",
       " SeqFeature(CompoundLocation([FeatureLocation(ExactPosition(8645), ExactPosition(8668), strand=1), FeatureLocation(ExactPosition(9382), ExactPosition(9431), strand=1)], 'join'), type='tRNA', location_operator='join'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(9589), ExactPosition(9661), strand=1), type='gene'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(9589), ExactPosition(9661), strand=1), type='tRNA'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(9937), ExactPosition(11461), strand=-1), type='gene'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(9937), ExactPosition(11461), strand=-1), type='CDS'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(11528), ExactPosition(12798), strand=-1), type='gene'),\n",
       " SeqFeature(CompoundLocation([FeatureLocation(ExactPosition(12653), ExactPosition(12798), strand=-1), FeatureLocation(ExactPosition(11528), ExactPosition(11938), strand=-1)], 'join'), type='CDS', location_operator='join'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(13261), ExactPosition(13507), strand=-1), type='gene'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(13261), ExactPosition(13507), strand=-1), type='CDS'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(14020), ExactPosition(14770), strand=-1), type='gene'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(14020), ExactPosition(14770), strand=-1), type='CDS'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(15012), ExactPosition(15723), strand=-1), type='gene'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(15012), ExactPosition(15723), strand=-1), type='CDS'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(15937), ExactPosition(20068), strand=-1), type='gene'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(15937), ExactPosition(20068), strand=-1), type='CDS'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(20250), ExactPosition(23084), strand=-1), type='gene'),\n",
       " SeqFeature(CompoundLocation([FeatureLocation(ExactPosition(22652), ExactPosition(23084), strand=-1), FeatureLocation(ExactPosition(20250), ExactPosition(21861), strand=-1)], 'join'), type='CDS', location_operator='join'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(23110), ExactPosition(26329), strand=-1), type='gene'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(23110), ExactPosition(26329), strand=-1), type='CDS'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(27372), ExactPosition(27443), strand=1), type='gene'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(27372), ExactPosition(27443), strand=1), type='tRNA'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(28088), ExactPosition(28178), strand=1), type='gene'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(28088), ExactPosition(28178), strand=1), type='CDS'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(28706), ExactPosition(28811), strand=-1), type='gene'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(28706), ExactPosition(28811), strand=-1), type='CDS'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(29800), ExactPosition(29874), strand=-1), type='gene'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(29800), ExactPosition(29874), strand=-1), type='tRNA'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(30322), ExactPosition(30406), strand=-1), type='gene'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(30322), ExactPosition(30406), strand=-1), type='tRNA'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(30465), ExactPosition(30538), strand=-1), type='gene'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(30465), ExactPosition(30538), strand=-1), type='tRNA'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(31368), ExactPosition(31440), strand=1), type='gene'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(31368), ExactPosition(31440), strand=1), type='tRNA'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(32710), ExactPosition(33772), strand=1), type='gene'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(32710), ExactPosition(33772), strand=1), type='CDS'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(33719), ExactPosition(35141), strand=1), type='gene'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(33719), ExactPosition(35141), strand=1), type='CDS'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(35311), ExactPosition(35403), strand=-1), type='gene'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(35311), ExactPosition(35403), strand=-1), type='tRNA'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(35750), ExactPosition(35939), strand=1), type='gene'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(35750), ExactPosition(35939), strand=1), type='CDS'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(36489), ExactPosition(36560), strand=1), type='gene'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(36489), ExactPosition(36560), strand=1), type='tRNA'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(36703), ExactPosition(36777), strand=-1), type='gene'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(36703), ExactPosition(36777), strand=-1), type='tRNA'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(36937), ExactPosition(37240), strand=-1), type='gene'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(36937), ExactPosition(37240), strand=-1), type='CDS'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(37374), ExactPosition(39579), strand=-1), type='gene'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(37374), ExactPosition(39579), strand=-1), type='CDS'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(39604), ExactPosition(41857), strand=-1), type='gene'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(39604), ExactPosition(41857), strand=-1), type='CDS'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(42583), ExactPosition(44591), strand=-1), type='gene'),\n",
       " SeqFeature(CompoundLocation([FeatureLocation(ExactPosition(44465), ExactPosition(44591), strand=-1), FeatureLocation(ExactPosition(43523), ExactPosition(43751), strand=-1), FeatureLocation(ExactPosition(42583), ExactPosition(42736), strand=-1)], 'join'), type='CDS', location_operator='join'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(44826), ExactPosition(44913), strand=1), type='gene'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(44826), ExactPosition(44913), strand=1), type='tRNA'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(45222), ExactPosition(45828), strand=-1), type='gene'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(45222), ExactPosition(45828), strand=-1), type='CDS'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(46212), ExactPosition(46285), strand=-1), type='gene'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(46212), ExactPosition(46285), strand=-1), type='tRNA'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(46893), ExactPosition(47490), strand=1), type='gene'),\n",
       " SeqFeature(CompoundLocation([FeatureLocation(ExactPosition(46893), ExactPosition(46928), strand=1), FeatureLocation(ExactPosition(47440), ExactPosition(47490), strand=1)], 'join'), type='tRNA', location_operator='join'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(48174), ExactPosition(48247), strand=1), type='gene'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(48174), ExactPosition(48247), strand=1), type='tRNA'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(48676), ExactPosition(49153), strand=-1), type='gene'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(48676), ExactPosition(49153), strand=-1), type='CDS'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(49256), ExactPosition(49934), strand=-1), type='gene'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(49256), ExactPosition(49934), strand=-1), type='CDS'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(50000), ExactPosition(50363), strand=-1), type='gene'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(50000), ExactPosition(50363), strand=-1), type='CDS'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(51198), ExactPosition(51871), strand=-1), type='gene'),\n",
       " SeqFeature(CompoundLocation([FeatureLocation(ExactPosition(51832), ExactPosition(51871), strand=-1), FeatureLocation(ExactPosition(51198), ExactPosition(51233), strand=-1)], 'join'), type='tRNA', location_operator='join'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(52055), ExactPosition(52128), strand=1), type='gene'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(52055), ExactPosition(52128), strand=1), type='tRNA'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(52264), ExactPosition(52663), strand=-1), type='gene'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(52264), ExactPosition(52663), strand=-1), type='CDS'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(52659), ExactPosition(54156), strand=-1), type='gene'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(52659), ExactPosition(54156), strand=-1), type='CDS'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(54957), ExactPosition(56397), strand=1), type='gene'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(54957), ExactPosition(56397), strand=1), type='CDS'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(57074), ExactPosition(58541), strand=1), type='gene'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(57074), ExactPosition(58541), strand=1), type='CDS'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(59246), ExactPosition(59360), strand=1), type='gene'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(59246), ExactPosition(59360), strand=1), type='CDS'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(59771), ExactPosition(60326), strand=1), type='gene'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(59771), ExactPosition(60326), strand=1), type='CDS'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(60740), ExactPosition(61430), strand=1), type='gene'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(60740), ExactPosition(61430), strand=1), type='CDS'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(61656), ExactPosition(62619), strand=1), type='gene'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(61656), ExactPosition(62619), strand=1), type='CDS'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(63537), ExactPosition(63660), strand=-1), type='gene'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(63537), ExactPosition(63660), strand=-1), type='CDS'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(63803), ExactPosition(63920), strand=-1), type='gene'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(63803), ExactPosition(63920), strand=-1), type='CDS'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(63941), ExactPosition(64061), strand=-1), type='gene'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(63941), ExactPosition(64061), strand=-1), type='CDS'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(64070), ExactPosition(64322), strand=-1), type='gene'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(64070), ExactPosition(64322), strand=-1), type='CDS'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(65711), ExactPosition(65807), strand=1), type='gene'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(65711), ExactPosition(65807), strand=1), type='CDS'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(65997), ExactPosition(66111), strand=1), type='gene'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(65997), ExactPosition(66111), strand=1), type='CDS'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(66228), ExactPosition(66302), strand=-1), type='gene'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(66228), ExactPosition(66302), strand=-1), type='tRNA'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(66489), ExactPosition(66563), strand=-1), type='gene'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(66489), ExactPosition(66563), strand=-1), type='tRNA'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(66928), ExactPosition(67063), strand=1), type='gene'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(66928), ExactPosition(67063), strand=1), type='CDS'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(67487), ExactPosition(67688), strand=1), type='gene'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(67487), ExactPosition(67688), strand=1), type='CDS'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(67916), ExactPosition(68222), strand=1), type='gene'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(67916), ExactPosition(68222), strand=1), type='CDS'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(68511), ExactPosition(68865), strand=-1), type='gene'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(68511), ExactPosition(68865), strand=-1), type='CDS'),\n",
       " SeqFeature(CompoundLocation([FeatureLocation(ExactPosition(69610), ExactPosition(69724), strand=-1), FeatureLocation(ExactPosition(139855), ExactPosition(140650), strand=1)], 'join'), type='gene', location_operator='join'),\n",
       " SeqFeature(CompoundLocation([FeatureLocation(ExactPosition(69610), ExactPosition(69724), strand=-1), FeatureLocation(ExactPosition(139855), ExactPosition(140087), strand=1), FeatureLocation(ExactPosition(140624), ExactPosition(140650), strand=1)], 'join'), type='CDS', location_operator='join'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(69909), ExactPosition(71882), strand=-1), type='gene'),\n",
       " SeqFeature(CompoundLocation([FeatureLocation(ExactPosition(71811), ExactPosition(71882), strand=-1), FeatureLocation(ExactPosition(70652), ExactPosition(70944), strand=-1), FeatureLocation(ExactPosition(69909), ExactPosition(70137), strand=-1)], 'join'), type='CDS', location_operator='join'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(72370), ExactPosition(73897), strand=1), type='gene'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(72370), ExactPosition(73897), strand=1), type='CDS'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(74081), ExactPosition(74183), strand=1), type='gene'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(74081), ExactPosition(74183), strand=1), type='CDS'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(74248), ExactPosition(74380), strand=-1), type='gene'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(74248), ExactPosition(74380), strand=-1), type='CDS'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(74484), ExactPosition(74706), strand=1), type='gene'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(74484), ExactPosition(74706), strand=1), type='CDS'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(74840), ExactPosition(76292), strand=1), type='gene'),\n",
       " SeqFeature(CompoundLocation([FeatureLocation(ExactPosition(74840), ExactPosition(74846), strand=1), FeatureLocation(ExactPosition(75650), ExactPosition(76292), strand=1)], 'join'), type='CDS', location_operator='join'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(76480), ExactPosition(77672), strand=1), type='gene'),\n",
       " SeqFeature(CompoundLocation([FeatureLocation(ExactPosition(76480), ExactPosition(76488), strand=1), FeatureLocation(ExactPosition(77197), ExactPosition(77672), strand=1)], 'join'), type='CDS', location_operator='join'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(77900), ExactPosition(78890), strand=-1), type='gene'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(77900), ExactPosition(78890), strand=-1), type='CDS'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(78959), ExactPosition(79376), strand=-1), type='gene'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(78959), ExactPosition(79376), strand=-1), type='CDS'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(79488), ExactPosition(79602), strand=-1), type='gene'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(79488), ExactPosition(79602), strand=-1), type='CDS'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(80067), ExactPosition(80472), strand=-1), type='gene'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(80067), ExactPosition(80472), strand=-1), type='CDS'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(80695), ExactPosition(81064), strand=-1), type='gene'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(80695), ExactPosition(81064), strand=-1), type='CDS'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(81188), ExactPosition(82652), strand=-1), type='gene'),\n",
       " SeqFeature(CompoundLocation([FeatureLocation(ExactPosition(82643), ExactPosition(82652), strand=-1), FeatureLocation(ExactPosition(81188), ExactPosition(81587), strand=-1)], 'join'), type='CDS', location_operator='join'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(82825), ExactPosition(83482), strand=-1), type='gene'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(82825), ExactPosition(83482), strand=-1), type='CDS'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(83466), ExactPosition(83949), strand=-1), type='gene'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(83466), ExactPosition(83949), strand=-1), type='CDS'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(84004), ExactPosition(84283), strand=-1), type='gene'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(84004), ExactPosition(84283), strand=-1), type='CDS'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(84336), ExactPosition(85843), strand=-1), type='gene'),\n",
       " SeqFeature(CompoundLocation([FeatureLocation(ExactPosition(85453), ExactPosition(85843), strand=-1), FeatureLocation(ExactPosition(84336), ExactPosition(84771), strand=-1)], 'join'), type='CDS', location_operator='join'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(85861), ExactPosition(86143), strand=-1), type='gene'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(85861), ExactPosition(86143), strand=-1), type='CDS'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(86311), ExactPosition(86385), strand=-1), type='gene'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(86311), ExactPosition(86385), strand=-1), type='tRNA'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(86473), ExactPosition(93358), strand=1), type='gene'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(86473), ExactPosition(93358), strand=1), type='CDS'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(94275), ExactPosition(94356), strand=-1), type='gene'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(94275), ExactPosition(94356), strand=-1), type='tRNA'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(94940), ExactPosition(97164), strand=-1), type='gene'),\n",
       " SeqFeature(CompoundLocation([FeatureLocation(ExactPosition(96387), ExactPosition(97164), strand=-1), FeatureLocation(ExactPosition(94940), ExactPosition(95702), strand=-1)], 'join'), type='CDS', location_operator='join'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(97477), ExactPosition(97945), strand=-1), type='gene'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(97477), ExactPosition(97945), strand=-1), type='CDS'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(100708), ExactPosition(100780), strand=1), type='gene'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(100708), ExactPosition(100780), strand=1), type='tRNA'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(101011), ExactPosition(102502), strand=1), type='gene'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(101011), ExactPosition(102502), strand=1), type='rRNA'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(102800), ExactPosition(103601), strand=1), type='gene'),\n",
       " SeqFeature(CompoundLocation([FeatureLocation(ExactPosition(102800), ExactPosition(102837), strand=1), FeatureLocation(ExactPosition(103566), ExactPosition(103601), strand=1)], 'join'), type='tRNA', location_operator='join'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(103664), ExactPosition(104538), strand=1), type='gene'),\n",
       " SeqFeature(CompoundLocation([FeatureLocation(ExactPosition(103664), ExactPosition(103702), strand=1), FeatureLocation(ExactPosition(104503), ExactPosition(104538), strand=1)], 'join'), type='tRNA', location_operator='join'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(104690), ExactPosition(107500), strand=1), type='gene'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(104690), ExactPosition(107500), strand=1), type='rRNA'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(107598), ExactPosition(107701), strand=1), type='gene'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(107598), ExactPosition(107701), strand=1), type='rRNA'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(108301), ExactPosition(108375), strand=1), type='gene'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(108301), ExactPosition(108375), strand=1), type='tRNA'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(109012), ExactPosition(109084), strand=-1), type='gene'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(109012), ExactPosition(109084), strand=-1), type='tRNA'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(109404), ExactPosition(110436), strand=1), type='gene'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(109404), ExactPosition(110436), strand=1), type='CDS'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(110397), ExactPosition(112638), strand=-1), type='gene'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(110397), ExactPosition(112638), strand=-1), type='CDS'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(113448), ExactPosition(113607), strand=1), type='gene'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(113448), ExactPosition(113607), strand=1), type='CDS'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(114269), ExactPosition(114349), strand=1), type='gene'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(114269), ExactPosition(114349), strand=1), type='tRNA'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(114460), ExactPosition(115447), strand=1), type='gene'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(114460), ExactPosition(115447), strand=1), type='CDS'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(115664), ExactPosition(117167), strand=-1), type='gene'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(115664), ExactPosition(117167), strand=-1), type='CDS'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(117317), ExactPosition(117563), strand=-1), type='gene'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(117317), ExactPosition(117563), strand=-1), type='CDS'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(117803), ExactPosition(118109), strand=-1), type='gene'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(117803), ExactPosition(118109), strand=-1), type='CDS'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(118376), ExactPosition(118907), strand=-1), type='gene'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(118376), ExactPosition(118907), strand=-1), type='CDS'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(119243), ExactPosition(119762), strand=-1), type='gene'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(119243), ExactPosition(119762), strand=-1), type='CDS'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(119846), ExactPosition(122009), strand=-1), type='gene'),\n",
       " SeqFeature(CompoundLocation([FeatureLocation(ExactPosition(121456), ExactPosition(122009), strand=-1), FeatureLocation(ExactPosition(119846), ExactPosition(120376), strand=-1)], 'join'), type='CDS', location_operator='join'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(122010), ExactPosition(123192), strand=-1), type='gene'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(122010), ExactPosition(123192), strand=-1), type='CDS'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(123295), ExactPosition(123562), strand=-1), type='gene'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(123295), ExactPosition(123562), strand=-1), type='CDS'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(123883), ExactPosition(129244), strand=-1), type='gene'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(123883), ExactPosition(129244), strand=-1), type='CDS'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(129564), ExactPosition(129636), strand=1), type='gene'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(129564), ExactPosition(129636), strand=1), type='tRNA'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(130273), ExactPosition(130347), strand=-1), type='gene'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(130273), ExactPosition(130347), strand=-1), type='tRNA'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(130579), ExactPosition(130700), strand=-1), type='gene'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(130579), ExactPosition(130700), strand=-1), type='rRNA'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(130947), ExactPosition(131050), strand=-1), type='gene'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(130947), ExactPosition(131050), strand=-1), type='rRNA'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(131148), ExactPosition(133958), strand=-1), type='gene'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(131148), ExactPosition(133958), strand=-1), type='rRNA'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(134110), ExactPosition(134984), strand=-1), type='gene'),\n",
       " SeqFeature(CompoundLocation([FeatureLocation(ExactPosition(134946), ExactPosition(134984), strand=-1), FeatureLocation(ExactPosition(134110), ExactPosition(134145), strand=-1)], 'join'), type='tRNA', location_operator='join'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(135047), ExactPosition(135848), strand=-1), type='gene'),\n",
       " SeqFeature(CompoundLocation([FeatureLocation(ExactPosition(135811), ExactPosition(135848), strand=-1), FeatureLocation(ExactPosition(135047), ExactPosition(135082), strand=-1)], 'join'), type='tRNA', location_operator='join'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(136146), ExactPosition(137637), strand=-1), type='gene'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(136146), ExactPosition(137637), strand=-1), type='rRNA'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(137868), ExactPosition(137940), strand=-1), type='gene'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(137868), ExactPosition(137940), strand=-1), type='tRNA'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(140703), ExactPosition(141171), strand=1), type='gene'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(140703), ExactPosition(141171), strand=1), type='CDS'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(141484), ExactPosition(143708), strand=1), type='gene'),\n",
       " SeqFeature(CompoundLocation([FeatureLocation(ExactPosition(141484), ExactPosition(142261), strand=1), FeatureLocation(ExactPosition(142946), ExactPosition(143708), strand=1)], 'join'), type='CDS', location_operator='join'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(144292), ExactPosition(144373), strand=1), type='gene'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(144292), ExactPosition(144373), strand=1), type='tRNA'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(145290), ExactPosition(152175), strand=-1), type='gene'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(145290), ExactPosition(152175), strand=-1), type='CDS'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(152263), ExactPosition(152337), strand=1), type='gene'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(152263), ExactPosition(152337), strand=1), type='tRNA'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(152505), ExactPosition(152787), strand=1), type='gene'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(152505), ExactPosition(152787), strand=1), type='CDS'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(152805), ExactPosition(154312), strand=1), type='gene'),\n",
       " SeqFeature(CompoundLocation([FeatureLocation(ExactPosition(152805), ExactPosition(153195), strand=1), FeatureLocation(ExactPosition(153877), ExactPosition(154312), strand=1)], 'join'), type='CDS', location_operator='join')]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "reference_record.features"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "fc583381",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index</th>\n",
       "      <th>qseqid</th>\n",
       "      <th>sseqid</th>\n",
       "      <th>pident</th>\n",
       "      <th>length</th>\n",
       "      <th>mismatch</th>\n",
       "      <th>gapopen</th>\n",
       "      <th>qstart</th>\n",
       "      <th>qend</th>\n",
       "      <th>sstart</th>\n",
       "      <th>send</th>\n",
       "      <th>evalue</th>\n",
       "      <th>bitscore</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>27</td>\n",
       "      <td>NC_000932.1</td>\n",
       "      <td>cds-blatx_matK_1;gbkey=CDS;gene=matK</td>\n",
       "      <td>100.0</td>\n",
       "      <td>1515</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2056</td>\n",
       "      <td>3570</td>\n",
       "      <td>1515</td>\n",
       "      <td>1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2914.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   index       qseqid                                sseqid  pident  length  \\\n",
       "2     27  NC_000932.1  cds-blatx_matK_1;gbkey=CDS;gene=matK   100.0    1515   \n",
       "\n",
       "   mismatch  gapopen  qstart  qend  sstart  send  evalue  bitscore  \n",
       "2         0        0    2056  3570    1515     1     0.0    2914.0  "
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dat [ dat[\"sseqid\"] == \"cds-blatx_matK_1;gbkey=CDS;gene=matK\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "4824b9db",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "SeqFeature(FeatureLocation(ExactPosition(382), ExactPosition(1444), strand=-1), type='CDS')"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "reference_record.features[6]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2602ba0f",
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio.SeqFeature import SeqFeature, FeatureLocation\n",
    "\n",
    "feature = SeqFeature(\n",
    "    location=FeatureLocation(2056, 3570, strand=-1), \n",
    "    type=\"CDS\",\n",
    "    id=\"matK\")\n",
    "feature.type = \"CDS\"\n",
    "\n",
    "query_record.features.append(feature)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "246b36a8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[SeqFeature(FeatureLocation(ExactPosition(2056), ExactPosition(3570), strand=-1), type='CDS', id='matK')]"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query_record.features"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "d0bcd9bc",
   "metadata": {},
   "outputs": [],
   "source": [
    "# remember to remove 10kb \n",
    "query_record.annotations[\"molecule_type\"] = \"DNA\"\n",
    "if query_circular: \n",
    "    query_record.annotations[\"topology\"] = \"circular\"\n",
    "else: \n",
    "    query_record.annotations[\"topology\"] = \"linear\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "e373eb25",
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import SeqIO\n",
    "\n",
    "with open(\"example.gb\", \"w\") as output_handle:\n",
    "    SeqIO.write(query_record, output_handle, \"gb\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "a3ba3c60",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "SeqRecord(seq=Seq('ATGGGCGAACGACGGGAATTGAACCCGCGATGGTGAATTCACAATCCACTGCCT...GCT'), id='NC_000932.1', name='NC_000932.1', description='NC_000932.1 Arabidopsis thaliana chloroplast, complete genome', dbxrefs=[])"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# read in query\n",
    "list(SeqIO.parse(\"query.fasta\", \"fasta\"))[0]"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
